README.md
SWAGGER and MOCK
================

### Swagger

Swagger is the tool selected to create the specification for each webservice using HTTP resources among hybris and the Single Page Application (SPA).

This tool has its own specification to define the services and can be seen in the [Github](https://github.com/wordnik/swagger-spec/blob/master/versions/1.2.md).

Also, to serve this specification (because it is web based) a NodeJS approach was selected because it is less resource consuming on the development machines and because its characteristics it has a simple framework to create the mocks based on the specification.
The framework used in NodeJS with ExpressJS was the [swagger-node-express](https://github.com/wordnik/swagger-node-express).

The front-end to display the specification is based on the [swagger-ui](https://github.com/wordnik/swagger-ui).

Basically, once you download the project the specs are inside the SPA_mock/specs directory, two files for each resource, one with the specification for each method and another one with the data and the code for the mock.
In case of any doubts on what you see in the swagger-ui (front-end) you can check inside the first file, for example, for authorizations resource, the first mentioned file is the authorizations.js and the second one is the authorizationsData.js.
Inside the first file, you will see a section called 'spec' where you will find the method specification in a JSON format based on the [JSON schema](http://json-schema.org/).

### Mock

The mock was created using the swagger-node-express framework and NodeJS + ExpressJS.

### Starting up the SPA_mock project

To see the specification or use the mock you need to:

Once you have downloaded the project from the SVN go to the SPA_mock directory and run:

    $ npm install
    $ node bin/www 
    Swagger and Mock listenning on http://localhost:3000

Now you are ready to see the specification and run the mock against that server and port.

### To request fixes or changes on the swagger/mock


CHANGES.md
CHANGES
=======
### Changes on 26-ago-2014

* /submit_order
    - Added itemPrice (numeric) at the line level and currency (string) at the header level.
    - Removed shippingMethod from response message

### Changes on 19-ago-2014

* GET, PUT /shopping_cart
    - Specification updated with shippingCondition

* /submit_order
    - Specification updated with shippingCondition

### Changes on 14-ago-2014

* /scheduled_items
    - Specification updated with shippingCondition field.

* /sold_to/ship_tos
    - Added shipTos structure. Mock updated accordingly.

### Changes on 11-ago-2014

* /products
    - Specification updated with productDivision field.

### Changes on 06-ago-2014

* POST, PUT /shopping_cart
    - Specification updated with length, grade, plant fields. Mock Updated accordingly.

### Changes on 04-ago-2014

* /sold_to/ship_tos
    - Change structure to old one

* /sold_to/ship_tos, /sold_to/ship_tos/ship_to_id
    - Update shippingMethod field to array of strings

* PUT /shopping_cart
    - Update of specification: new field added: shipping_condition

* /statements
    - Mock updated with Account Summary data.
    New scenarios - 3 currencies - as of date: 06/10/2014, 04/09/2014, 04/11/2014, 03/09/2014, 03/10/2014, 03/11/2014
    New scenarios - 2 currencies - as of date: 05/09/2014, 05/10/2014, 05/11/2014, 06/09/2014, 06/11/2014
    New scenarios - 1 currencies - as of date: 04/10/2014

### Changes on 01-ago-2014

* /scheduled_items, /order_items
	- Structures updated as IDC observations

### Changes on 31-jul-2014

* /scheduled_items
    - Updated shipto structure. Mock updated accordingly.

* /sold_to/ship_tos
    - Added shipTos structure. Mock updated accordingly.

* /order_items
    - Changed shipto structure. Mock updated accordingly.


### Changes on 27-jul-2014
     /shipment_document
     	-Mock updated

     /schuedueld_items
	-Mock updated (scenario: P.O.=1)

### Changes on 18-jul-2014

* /saved_searches
    - gridConfig field was removed from structure. Mock updated accordingly.

* /statements
    - Mock scenarios updated due to format validate defined (Valid scenarios: 03/09/2014, 03/10/2014, 03/11/2014, 04/09/2014, 04/10/2014, 04/11/2014, 05/09/2014, 04/10/2014, 04/11/2014, 06/09/2014, 06/10/2014, 06/11/2014)


### Changes on 17-jul-2014

* /statements
    - Fix: Added missing fields urlInvoiceDebits and urlCreditsMemo in structure Debit and Credit respectively.  Mock updated accordingly.
    - Data added on mock (Valid scenarios: 2014-09-03, 2014-10-03, 2014-11-03, 2014-09-04, 2014-10-04, 2014-11-04, 2014-09-05, 2014-10-05, 2014-11-05, 2014-09-06, 2014-10-06, 2014-11-06)


### Changes on 15-jul-2014

* /shopping_cart
    - preferredQuantity field on RoundedError structure updated with lowercase data (up, down)

### Changes on 14-jul-2014

* /profile/dateFmt
    - New service. Mock created accordingly.

### Changes on 11-jul-2014

* /products
    - Fix on the mock. Body request contains field shapesSizesIds.

### Changes on 10-jul-2014

* /order_items
    - cwtPrice field updated to String type
    - Mock updated accordingly

### Changes on 08-jul-2014

* /scheduled_items
    - Mock updated with shape groups, plant and ship to filters

* /railcar_tracking
    - Railcar structure updated.
    - Mock updated accordingly. (valid scenarios - shippedDate: 2014-10-05 or 2014-05-05; bolNumber: 1 or 2; vehicleNumber: 1 or 2)

* /ship_tos
    -Mock fixed.

* /submit_order
    - Structure fixed (one character starting in UpperCase)
    - Mock fixed

* /saved_searches
    - bug fixes on the code that didn't change any service behavior.

* /shape_groups, /shopping_cart, /submit_order, /products, /order_items
    - sold to id


### Changes on 07-jul-2014

* /emails
    - Mock fixed

* /scheduled_items
    - Heat structure created
    - Init and end dates not mandatory
    - Mock updated

* /ship_tos
    - Mock updated to shipTos

* /shopping_cart
    - Items from ShoppingCartNewItem model corrected

* /statements
    - Mock updated (asOfDate: 2014-10-04, 2014-10-06)
    - Mock fixed

### Changes on 03-jul-2014

* all services
    - Added HTTP code 401 and 500.

* /emails
    - Request updated

* /downloads
    - Mock updated

* /scheduled_items
    - heatNumber updated to array of strings

* /shopping_cart
    - Mock fixed

* /statements
    - originalAmount field from Credit structure removed.
    - Mock updated

* /submit_order
    - Mock fixed

### Changes on 02-jul-2014

* /products
    - request changed to be compliant with terms used for the parameters like on version 1.13
    - quantity was removed from the response

* /order_items
    - orderQuantityUOM and shippedQuantityUOM changed to string instead of number

* /authorizations
    - Added HTTP code 400.
    - Mock updated.

* /contacts
    - Added HTTP code 400.

*/downloads
    - Added HTTP code 400.
    - Added response structure

* /emails
    - Added HTTP code 400.

* /family_sizes
    - Added HTTP code 400.

* /keep_alive
    - Added HTTP code 400.

* /order_items
    - Added HTTP code 400.
    - Mock updated (Error scenario changed to 400)

* /railcar_tracking
    - Added HTTP code 400.
    - Mock updated (Error scenario changed to 400)

* /saved_searches
    - Added HTTP code 400.

* /scheduled_items
    - Added HTTP code 400.
    - Mock updated (Error scenario changed to 400)

* /shape_groups
    - Added HTTP code 400.

* /shipment_documents
    - Added HTTP code 400.
    - Mock updated (Error scenario changed to 400)

* GET /shopping_cart
    - Added HTTP code 400.
    - shoppingCartItemsDesc changed from array to a simple object

* POST PUT /shopping_cart
    - Mock updated (Error scenario changed to 400)
    - poNumber and requestedDeliveryDate removed from the request message (only for the POST method)

* /sold_to, /ship_tos, /ship_to
    - Added HTTP code 400.
    - Mock updated (Error scenario changed to 400)

* /statements
    - Added HTTP code 400.

specs/models.js
exports.models = {
    "Error": {
        "id": "Error",
        "properties": {
            "severity": {
                "required": true,
                "type": "string",
                "enum": [
                    "Error",
                    "Warning"
                ]
            },
            "code": {
                "required": true,
                "type": "string"
            },
            "parameters": {
                "required": true,
                "type": "array",
                "items":
                {
                    type: 'string'
                }
            }
        }
    }
    ,
    "MyStartDTO":{
        "copyFlag": {
            "required": true,
            "type": "boolean",
            "description": "copyFlag"
        },
        "myStartName": {
            "required": true,
            "type": "string",
            "description": "myStartName"
        }
        ,
        "subscribed": {
            "required": true,
            "type": "string",
            "description": "subscribed"
        }
        ,
        "listType": {
            "required": true,
            "type": "integer",
            "description": "listType"
        },
        "noOfSubscribers": {
            "required": true,
            "type": "integer",
            "description": "myStartName"
        },
        "myStartId": {
            "required": true,
            "type": "integer",
            "description": "myStartName"
        },
        "myStartType": {
            "required": true,
            "type": "string",
            "description": "myStartType"
        },
        "published": {
            "required": true,
            "type": "string",
            "description": "published"
        },
        "userId": {
            "required": true,
            "type": "string",
            "description": "userId"
        },
        "msorId": {
            "required": true,
            "type": "integer",
            "description": "msorId"
        },
        "sequence": {
            "required": true,
            "type": "integer",
            "description": "sequence"
        }

    }
};

specs/datapool.js
var sw = require("swagger-node-express");
const delay = require('delay');
var poolData = require("./poolData.js");

// the description will be picked up in the resource listing
exports.getDataPool = {
    'spec': {
        description : "Operations about authorization",
        path : "/poolData",
        method: "GET",
        summary : "Retrieves pool data from backend.",
        notes : "This service is related to pool the data after time out from server.",
        type : "poolData",
        nickname : "getPoolData",
        produces : ["application/json"],
        parameters : [
            {
                "name": "request_id",
                "description": "Page where attributes will be loaded",
                "required": true,
                "type": "string",
                "paramType": "query"
            }
        ],
        responseMessages : [
            {
                code: 200,
                message: "OK"
            },
            {
                code: 400,
                message: "Bad request"
            },
            {
                code: 401,
                message: "Unauthorized"
            },
            {
                "code": 404,
                "message": "Not found."
            },
            {
                code: 500,
                message: "Internal Server Error"
            }
        ]
    },
    'action': function (req,res) {
        delay(20)
            .then(() => {
            // Executed after 200 milliseconds
            console.log("werrtttttttttttttttttttt");

        var entry = req.query.request_id;
        var output = poolData.getpoolData(entry);
        if (!output){
            //error
            var error ={} ;
            error.severity = 'Error';
            error.code = 'E-X';
            error.parameters=['Unauthorized'];

            output = {};
            output.authorizations = [];
            output.errors = error;

            res.json(401, output);
        } else {
            res.json(output);
        }
    });
    }
};


specs/attachments.js
var sw = require("swagger-node-express");

var attachmentsData = require("./attachmentsData.js");

exports.getAttachments = {
    'spec': {
        path : "/complaints/complaintDetails",
        method: "GET",
        summary : "POST documents attached",
		notes : "",
        type : "ResponseAttachments",
        nickname : "getAttachments",
        parameters : [
            {
                "name": "complaintId",
                "description": "Filters the documents attached search.",
                "required": false,
                "type": "FilterAttachments",
                "paramType": "query"
            }
        ],
        responseMessages : [
            {
                code: 200,
                message: "OK"
            },
            {
                code: 400,
                message: "Bad request"
            },
            {
                code: 401,
                message: "Unauthorized"
            },
            {
                "code": 404,
                "message": "Not found."
            },
            {
                code: 500,
                message: "Internal Server Error"
            }
        ]
    },
    'action': function (req,res) {
        //get body
        var body = req.query;
		        console.log(body);

        var output = attachmentsData.getAttachments(body);
        var outputError = [];

        var response = {
            attachmentList: output,
            errors: outputError
        };

        res.json(response);
    }
};


specs/authorizations.js
var sw = require("swagger-node-express");
const delay = require('delay');
var authorizationsData = require("./authorizationsData.js");

// the description will be picked up in the resource listing
exports.getAuthorizations = {
    'spec': {
        description : "Operations about authorization",
        path : "/BusinessFoundation-WS/rest/mystart/getmystartlist",
        method: "GET",
        summary : "Retrieves visibility rules.",
        notes : "This service is related to components visibility on the UI. It is important to notice that invisible components should not have returned information from hybris server on its specific integration, it means that UI is just hiding fields, the information is not there anyway.",
        type : "ResponseAuthorizations",
        nickname : "getAuthorizations",
        produces : ["application/json"],
        parameters : [
            {
                "name": "page_id",
                "description": "Page where attributes will be loaded",
                "required": true,
                "type": "string",
                "paramType": "query"
            }
        ],
        responseMessages : [
            {
                code: 200,
                message: "OK"
            },
            {
                code: 400,
                message: "Bad request"
            },
            {
                code: 401,
                message: "Unauthorized"
            },
            {
                "code": 404,
                "message": "Not found."
            },
            {
                code: 500,
                message: "Internal Server Error"
            }
        ]
    },
    'action': function (req,res) {
        delay(20000)
            .then(() => {
            // Executed after 200 milliseconds
            console.log("werrtttttttttttttttttttt");

        var entry = req.query.page_id;
        var output = authorizationsData.getAuthorizations(entry);
        if (!output){
            //error
            var error ={} ;
            error.severity = 'Error';
            error.code = 'E-X';
            error.parameters=['Unauthorized'];

            output = {};
            output.authorizations = [];
            output.errors = error;

            res.json(401, output);
        } else {
            res.json(output);
        }
    });
    }
};


app.js
var express = require('express');
var http = require('http');
var path = require('path');
var favicon = require('static-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');

var routes = require('./routes');
var users = require('./routes/user');

var app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(favicon());
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded());
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));
app.use(app.router);

app.get('/', routes.index);
app.get('/users', users.list);

/// catch 404 and forwarding to error handler
app.use(function(req, res, next) {
    var err = new Error('Not Found');
    err.status = 404;
    next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
    app.use(function(err, req, res, next) {
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
    res.render('error', {
        message: err.message,
        error: {}
    });
});

/*
 * ## Swagger - setup
 */

app.all('/*', function(req, res, next) {
    res.header("Access-Control-Allow-Origin", "*");
    res.header("Access-Control-Allow-Methods", "GET, PUT, POST, DELETE, OPTIONS");
    res.header("Access-Control-Allow-Headers", "X-Requested-With, Content-Type");
    next();
});

var swagger = require('swagger-node-express').createNew(app);

swagger.setAppHandler(app);

// This is a sample validator. It simply says that for _all_ POST, DELETE, PUT
// methods, the header `api_key` OR query param `api_key` must be equal
// to the string literal `special-key`. All other HTTP ops are A-OK
//swagger.addValidator(
//    function validate(req, path, httpMethod) {
//        // example, only allow POST for api_key="special-key"
//        if ("POST" == httpMethod || "DELETE" == httpMethod || "PUT" == httpMethod) {
//            var apiKey = req.headers["api_key"];
//            if (!apiKey) {
//                apiKey = url.parse(req.url,true).query["api_key"]; }
//            if ("special-key" == apiKey) {
//                return true;
//            }
//            return false;
//        }
//        return true;
//    }
//);

var models = require("./specs/models.js");

var authorizationsResources     = require("./specs/authorizations.js");
var poolData     = require("./specs/datapool.js");

// Add models and methods to swagger
swagger.addModels(models)
   // .addGet(authorizationsResources.getAuthorizations)
    .addGet(poolData.getDataPool)
//swagger.configureDeclaration("pet", {
//    description : "Operations about Pets",
//    authorizations : [],
//    produces: ["application/json"]
//});

// set api info
swagger.setApiInfo({
    title: "Start- HTTP webservices specification",
    "description": "<p>Start has heavy use of JavaScripts (JS). Single Page Application (SPA) approach is the architectural approach to improve the user experience (UX).</p>",
    termsOfServiceUrl: "",
    contact: "subash.rout@capgemini.com",
    license: "",
    licenseUrl: ""
});

//swagger.setAuthorizations({
//    apiKey: {
//        type: "apiKey",
//        passAs: "header"
//    }
//});

// Configures the app's base path and api version.
swagger.configureSwaggerPaths("", "api-docs", "");
swagger.configure("http://localhost:3000", "1.0.0");

// Serve up swagger ui at /docs via static route
var docs_handler = express.static(__dirname + '/ui/');
app.get(/^\/docs(\/.*)?$/, function(req, res, next) {
    if (req.url === '/docs') { // express static barfs on root url w/o trailing slash
        res.writeHead(302, { 'Location' : req.url + '/' });
        res.end();
        return;
    }
    // take off leading /docs so that connect locates file correctly
    req.url = req.url.substr('/docs'.length);
    return docs_handler(req, res, next);
});


module.exports = app;
